sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -352,22 +353,35 @@ class CacheManager extends Logging with AdaptiveSparkPlanHelper {
     }
     needToRecache.foreach { cd =>
       cd.cachedRepresentation.cacheBuilder.clearCache()
+      tryRebuildCacheEntry(spark, cd).foreach { entry =>
+        this.synchronized {
+          if (lookupCachedDataInternal(entry.plan).nonEmpty) {
+            logWarning("While recaching, data was already added to cache.")
+          } else {
+            cachedData = entry +: cachedData
+            CacheManager.logCacheOperation(log"Re-cached Dataframe cache entry:" +
+              log"${MDC(DATAFRAME_CACHE_ENTRY, entry)}")
+          }
+        }
+      }
+    }
+  }
+
+  private def tryRebuildCacheEntry(
+      spark: SparkSession,
+      cd: CachedData): Option[CachedData] = {
+    try {
       val sessionWithConfigsOff = getOrCloneSessionWithConfigsOff(spark)
       val (newKey, newCache) = sessionWithConfigsOff.withActive {
         val refreshedPlan = V2TableRefreshUtil.refresh(cd.plan)
```

Comment:
If we want to avoid failure here, how about skip query plan validation in the method call here instead?
```
val refreshedPlan = V2TableRefreshUtil.refresh(cd.plan, validation = false)
```
and skip the following in `V2TableRefreshUtil.refresh`
```
        validateTableIdentity(currentTable, r)
        validateDataColumns(currentTable, r)
        validateMetadataColumns(currentTable, r)
```